---
output:
  pdf_document: default
  html_document: default
---
# Pathway and Gene Ontology Enrichment

## Cytoscape and ClueGO

The ClueGO plugin of Cytoscape can run both **pathway enrichment** and **ontology enrichment** analysis. In other words, we can check which pathways and ontologies are overrepresented by our set of interesting genes (for example, differentially expressed genes or genes from interesting expression modules). And we can see if any of those pathways and ontologies contain more genes of interest than we would expect if we had instead grabbed the same number of random genes to feed into the enrichment analysis.

Here we will use the KEGG (Kyoto Encyclopedia of Genes and Genomes) PATHWAY database (https://www.genome.jp/kegg/pathway.html) to look for enriched pathways. And we will use GO (gene ontology) ontologies (https://geneontology.org/) to look for enriched ontologies. Feel free to try other pathway and ontology options.

**KEGG PATHWAY**

Pathways for:  
+ Metabolism  
+ Genetic Information Processing  
+ Cellular Processes  
+ Organismal Systems  
+ Human Diseases  
+ Drug Development


**Gene Ontology**

Gene ontologies describe the function of genes in biological systems using a structured vocabulary that crosses species boundaries and that is both human-readable and machine-readable. Common gene ontology sets include:  
+ Biological Process  
+ Molecular Function  
+ Cellular Component


Open the Cytoscape app, and load the ClueGO plug-in: Apps --> App Manager

Change the species to mouse.

Open the results table, copy the DEGs, and paste into clueGO.

Select in the "Visual Style" box sets the factors that will determine node properties. Keep the default selection, "Groups".

Pick ontologies and set differing node shapes.

Clicking on the "Evidence" button will display a decision tree clarifying evidence codes. Keep "All" checked for now.

Check the "Show only Pathways with pV <=" box and change it to 0.01. If you want the details on how p-values are calculated for GO terms, click the “?” icon.

Expand the "Statistical Options" menu. In the first box, change the selection from the default ("Enrichment/Depletion") to "Enrichment.

Run the analysis! It will take several minutes. When it’s finished, look to the bottom right of the screen. Select the "ClueGO Results (Cluster #1)" tab, and click the indicated icon to save ClueGO results tables to your machine.

Open the new directory. Among the several files, we can find a .png with GO terms represented with a horizontal bar graph as they are shown in the “Cluster #1” tab of in the software. Adjacent bars of the same color are terms that belong to the same group.

The most generally useful file is “NodeAttributeTables.txt”. The fields here indicate annotation source, p-values, and gene composition and for each gene. Note the "GOLevels” column and the “Overview Term” column on the far right. Let’s return to Cytoscape to clarify what these mean.

Each GO term has one or more associated levels. Only those with a level within the values in the “GO Tree Interval” have the potential to appear in the results. The boundaries of this window can be changed directly, or by dragging the “Network Specificity” bar.

## Additional options

Let's inspect some additional options.

It’s typically a good idea to check the “Use GO Term Fusion” box. This will improve runtime and simplify the results by eliminating redundancy.

Below the “GO Tree Interval” box, there are two others. The “GO Term/Pathway Selection” box will set the criterion for what is a qualifying GO term.

The “GO Term/Pathway Network Connectivity” scroll tab will set the kappa score threshold for what the edges that appear between nodes. Kappa score is a measure of relation any and all GO terms based on the number of associated genes they have in common. For more details regarding this calculation, click on the “?” icon.

Expand the “Grouping Options” box. Recall the variable (True/False) of the last column (Overview Term) in the node attribute table file we looked at. Those with a “True” value (as well as a check mark next to the GO id in the “Clue GO Results” tab in Cytoscape) are the leading term in the group, only meaning it is the group with the lowest p-value.

The criterion to determine the leading term can be changed via the “Leading Group Term based on” selection. The threshold of commonality for what constitutes a group can also be adjusted, if so chosen.

Let's run the analysis again, selecting the KEGG ontology instead of the GO ontologies.

## CluePedia

Select the "CluePedia" tab, and click the network icon. This will add miniature gene nodes within the edges between nodes. As of now, however, it is not apparent which of them are upregulated and which are downregulated.

Navigate to File –> Import –> Table from File, and select the document from which the gene list was originally input. In the new window, chose “To selected networks only”. Under the “Key Column for Network” menu, pick “Input Gene ID”, then “OK”.

In the control panel, on the left side of Cytoscape, select ”Style”. For column, choose “Log2Fold change” and for mapping type, select “Continuous Mapping” and click the spectrum map.

In the window that appears, drag the handle position to 0 and change the color to grey or white. Then, drag the margins in to about a log2FoldChange of 2 and change their color. Change the arrows on the extremes to the same color.

You can save the session by clicking the icon at the top right of the app, which will allow you to return to the visualization and settings of the analysis in another session.

## Revigo

Revigo (reduce & visualize gene ontology) is a tool that will collapse related GO terms and create some interactive visualizations.

Copy your enriched GO terms ("ID" column) and the "Term PValue Corrected with Bonferroni step down" column (note: it is ok to include the KEGG pathways that are in the "ID" column, Revigo will just warn you that it can't find any matches for them and will ignore them).

Go to http://revigo.irb.hr/ and paste them in the box.

Change the species to "Homo sapiens".

Click on "Start Revigo".

After we explore the results, try rerunning it, changing "How large would you like the resulting list to be" to different values.
