<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <title>Differential Expression Workshop NCGR</title>

  
   <meta name="description" content="<p>This is the material for a Differential Gene Expression workshop offered by NCGR in collaboration
with NM-INBRE.</p>" />
   <meta name="generator" content="placeholder" />
  <meta property="og:title" content="Differential Expression Workshop NCGR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="<p>This is the material for a Differential Gene Expression workshop offered by NCGR in collaboration
with NM-INBRE.</p>" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Differential Expression Workshop NCGR" />
  
  <meta name="twitter:description" content="<p>This is the material for a Differential Gene Expression workshop offered by NCGR in collaboration
with NM-INBRE.</p>" />
  
  <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script>
  <script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script>
    <script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
    <link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet" />
    <script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script>
    <script src="libs/bs3compat-0.4.2/transition.js"></script>
    <script src="libs/bs3compat-0.4.2/tabs.js"></script>
    <script src="libs/bs3compat-0.4.2/bs3compat.js"></script>
    <link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet" />
    <script src="libs/bs4_book-1.0.0/bs4_book.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script>

  <!-- CSS -->
  <style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
    
</head>

<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<!--bookdown:title:start-->
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book">
    <a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Differential Expression Workshop NCGR</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
      </form>

      <nav aria-label="Table of contents">
        <h2>Table of contents</h2>
        <div id="book-toc"></div>

        <div class="book-extra">
          <p><a id="book-repo" href="#">View book source <i class="fab fa-github"></i></a></li></p>
        </div>
      </nav>
    </div>
  </header>

  <main class="col-sm-12 col-md-9 col-lg-7" id="content">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="national-center-for-genome-resources" class="section level1 unnumbered">
<h1 class="unnumbered">National Center for Genome Resources</h1>
<p>This publication was supported by an Institutional Development Award (IDeA) from the National Institute of General Medical Sciences of the National Institutes of Health under grant number P20GM103451.</p>
<div id="license-and-copyright" class="section level2 unnumbered">
<h2 class="unnumbered">License and Copyright</h2>
<p>Creative Commons Attribution-NonCommercial-NoDerivatives 4.0
<a href="https://creativecommons.org/licenses/by-nc-nd/4.0/" class="uri">https://creativecommons.org/licenses/by-nc-nd/4.0/</a></p>
<p>© 2023 National Center for Genome Resources</p>
<!--chapter:end:index.Rmd-->
</div>
</div>
<div id="de-chapter-2" class="section level1" number="1">
<h1><span class="header-section-number">1</span> DE Chapter 2</h1>
<!--chapter:end:02.Rmd-->
</div>
<div id="sequence-metrics-and-quality-control" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Sequence metrics and quality Control</h1>
<div id="the-study-design" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> The Study Design</h2>
<p>3 replicates of control mice and 3 replicates of water avoidance stress induced mice. We will run our own analysis on these publically-available data and compare the DEG and pathway enrichment results to those published.
<a href="https://www.mdpi.com/2218-1989/13/3/453" class="uri">https://www.mdpi.com/2218-1989/13/3/453</a></p>
</div>
<div id="data-access" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Data Access</h2>
<p>Start a screen, make a new directory branch “DGE_workshop/reads”, and enter it. Though there are only 6 samples, the sequencing depth is very high, making the download time ~ 1hr. If time does not allow for this (Option 1), cancel the command after checking it works, and softlink the data instead (Option 2).</p>
<p>OPTION 1:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate fastqdump</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">line</span><span class="kw">;</span> <span class="cf">do</span> <span class="ex">fastq-dump</span> <span class="at">--gzip</span> <span class="at">--outdir</span> ./ <span class="va">${line}</span><span class="kw">;</span> <span class="cf">done</span> <span class="op">&lt;</span> /home/elavelle/SRRlist.txt</span></code></pre></div>
<p>OPTION 2:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> home/elavelle/DGE_workshop/reads<span class="pp">*</span> ./</span></code></pre></div>
</div>
<div id="fastp-qc" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> fastp QC</h2>
<p>Takes ~ 15 min. for all 6 samples on 10 threads.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate fastp</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> SRR<span class="pp">*</span><span class="kw">;</span> <span class="cf">do</span> <span class="ex">fastp</span> <span class="at">-w</span> 10 <span class="at">-i</span> <span class="va">${file}</span> <span class="at">-o</span> <span class="va">${file}</span>.fastp.gz<span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p>This sort of messes up the extensions. Rename the files</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rename</span> .gz.fastp.gz .fastp.gz <span class="pp">*</span>.gz</span></code></pre></div>
<p>Download and view the .html report.</p>
</div>
<div id="checking-sequence-metrics" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Checking Sequence Metrics</h2>
<p>awk “NR” variable: There are several built-in awk variables. One of them is “NR” (Number of Rows/
Records). It contains line number and can be used to determine the total number of lines in a file. The
following example illustrates one use of this built-in variable.</p>
<p>Prints the line number of the first 10 lines</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">zcat</span> SRR23869771.fastq.fastq.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{print NR}&#39;</span> <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<p>awk modulo arithmatic: The modulo operator, “%”, returns the remainder of division. The expression
“5 % 2”, for example, would evaluate to 1, while “9 % 3” would evaluate to 0 because the dividend (9) is
a multiple of the divisor (3).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">zcat</span> SRR23869771.fastq.fastq.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{ if ( NR % 4 == 2) print length ($1)}&#39;</span> <span class="kw">|</span> <span class="fu">head</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">zcat</span> SRR23869771.fastq.fastp.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{ if ( NR % 4 == 2) print $0}&#39;</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<!--chapter:end:03QC.Rmd-->
</div>
</div>
<div id="de-chapter-4" class="section level1" number="3">
<h1><span class="header-section-number">3</span> DE Chapter 4</h1>
<!--chapter:end:04.Rmd-->
</div>
<div id="read-alignment" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Read Alignment</h1>
<div id="reference-acquisition" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Reference acquisition</h2>
<p>Create a new directory under DGE_workshop</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> genome</span></code></pre></div>
<p>From this page: <a href="https://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Mus_musculus/latest_assembly_versions/GCF_000001635.27_GRCm39/" class="uri">https://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Mus_musculus/latest_assembly_versions/GCF_000001635.27_GRCm39/</a>
Download the appropriate genome and annotation.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Mus_musculus/latest_assembly_versions/GCF_000001635.27_GRCm39/GCF_000001635.27_GRCm39_genomic.fna.gz</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Mus_musculus/latest_assembly_versions/GCF_000001635.27_GRCm39/GCF_000001635.27_GRCm39_genomic.gff.gz</span></code></pre></div>
</div>
<div id="hisat2" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> HISAT2</h2>
<p>Activate environment, and use HISAT2 to create indices. If time prohibits, we will likely need to terminate and softlink to output files.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate HISAT</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="ex">hisat2</span> <span class="at">-x</span> /home/<span class="va">$USER</span>/DGE_workshop/genome/GRCm39.index <span class="at">--threads</span> 10 <span class="at">-U</span> read <span class="at">-S</span> <span class="va">${line}</span>.sam<span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p>Run alignments</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> <span class="pp">*</span>.fastp.gz<span class="kw">;</span> <span class="cf">do</span> <span class="ex">hisat2</span> <span class="at">-x</span>/home/<span class="va">$USER</span>/DGE_workshop/genome/GRCm39.index <span class="at">--threads</span> 10 <span class="at">-U</span> <span class="va">${file}</span> <span class="at">-S</span> /home/<span class="va">$USER</span>/DGE_workshop/alignments/<span class="va">${file}</span>.sam<span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p>What flags would you use for paired-end reads?</p>
</div>
<div id="alignment-metrics" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Alignment metrics</h2>
<p>Some alignment tools (HISAT2 for example) will print alignment metrics after generating alignments. However, these metrics may not be available as a result of other alignment tools. Hence, it is useful to know the following one-liners to extract information on important metrics from SAM files.</p>
<p>Change environments and use this statistics summarization command. It takes about a minute per sample.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> flagstat SRR23869771.fastq.fastp.gz.sam</span></code></pre></div>
<p>The number in the first row is the count of records in the .sam file. Confirm this by counting the rows not including the header lines.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;^@&quot;</span> SRR23869771.fastq.fastp.gz.sam</span></code></pre></div>
<p>Secondary alignments counts the alignments of reads that mapped to additional locations on the genome. To omit these, Count the number of unique read IDs in the file.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;^@&quot;</span> SRR23869771.fastq.fastp.gz.sam <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{print $1}&#39;</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<p>Check this with arithmatic from the flagstat output: 38991995 - 14794600</p>
<p>However, this count also includes reads which didn’t map at all. Filter those out (“*” in column 3) to find the number of reads that mapped once or more.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> SRR23869771.fastq.fastp.gz.sam <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&#39;^@&#39;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{ if ($3 != &quot;*&quot;) print $0}&#39;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{print $1}&#39;</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<p>The ” mapped” number from the flagstat output counts all alignments (not reads!) Subtract the secondary alignments from that value to check our result:
35942177 - 14794600</p>
<p>As it happens, there is another convenient samtools function to extract the desired metrics from # a .sam file.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-f</span> 0x100 <span class="at">-c</span> SRR23869771.fastq.fastp.gz.sam </span></code></pre></div>
<p>Including the “-f” option in the samtools “view” command will print to stdout the records matching the corresponding bit flag shown in the table below. This command, for example, counts (due to the inclusion of the “-c” option) all secondary reads. Notice it matches the number found with the other methods.</p>
<p>The “-F” option is similar to grep’s “-v” option; it pulls the OPPOSITE records from what the bit flag describes. Moreover, these bit flags can be combined- e.g., 904 = 800 + 100 + 4
Therefore, the number of primary alignments can also be found by:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-F</span> 0x904 <span class="at">-c</span> SRR23869771.fastq.fastp.gz.sam </span></code></pre></div>
<p>Exercise: Select another sample, and using any of the means above, determine 1) Total read count 2) Total mapped reads 3) Total primary alignments</p>
</div>
<div id="sam.bam-conversion-and-alignment-sorting" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> .sam/.bam conversion and alignment sorting</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> <span class="pp">*</span>.sam<span class="kw">;</span> <span class="cf">do</span> <span class="ex">samtools</span> view <span class="at">-u</span> <span class="va">${file}</span> <span class="kw">|</span> <span class="ex">samtools</span> sort <span class="at">-o</span> <span class="va">${file}</span>.sorted<span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p>Rename the output files</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rename</span> .fastq.fastp.gz.sam.sorted .sorted.bam <span class="pp">*</span>.sorted</span></code></pre></div>
<p>Exercise: What average genome coverage do our primary alignments give us?</p>
<!--chapter:end:05Alignment.Rmd-->
</div>
</div>
<div id="abundance-estimation" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Abundance Estimation</h1>
<div id="format-check" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Format check</h2>
<p>Look at the format for a .gff file: <a href="https://en.wikipedia.org/wiki/General_feature_format" class="uri">https://en.wikipedia.org/wiki/General_feature_format</a></p>
<p>Exercise: Which genome/annotation pair in the folder /home/elavelle/DGE_Virtual/exercise_genomes/ has mismatched headers?</p>
</div>
<div id="featurecounts" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> FeatureCounts</h2>
<p>Create a new directory under DGE_workshop</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate featurecounts</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="ex">featureCounts</span> <span class="at">-a</span> /home/<span class="va">$USER</span>/DGE_workshop/genome/GRCm39.index <span class="at">-o</span> /home/<span class="va">$USER</span>/DGE_workshop/count_matrix.tsv <span class="at">--largestOverlap</span> <span class="at">-M</span> <span class="at">--primary</span> <span class="at">-T</span> 10 <span class="at">-t</span> gene <span class="at">-g</span> gene_name <span class="pp">*</span>.bam</span></code></pre></div>
<p>FeatureCounts put in an extra row and some columns we want to get rid of before doing differential expression analysis.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> count_matrix.tsv</span></code></pre></div>
<p>Use tail to take every row starting with the second, then extract only the columns of interest.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> +2 count_matrix.tsv <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{print $1 &quot;\t&quot; $7 &quot;\t&quot; $8 &quot;\t&quot; $9 &quot;\t&quot; $10 &quot;\t&quot; $11 &quot;\t&quot; $12 &quot;\t&quot; $13 &quot;\t&quot; $14 &quot;\t&quot; $15 &quot;\t&quot; $16 &quot;\t&quot; $17 &quot;\t&quot; $18}&#39;</span> <span class="op">&gt;</span> DESEQ2_matrix.tsv</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> /home/<span class="va">$USER</span>/DGE_Virtual/DESeq2</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> DESEQ2_matrix.tsv /home/<span class="va">$USER</span>/DGE_Virtual/DESEQ2/</span></code></pre></div>
<!--chapter:end:06Abundancy_Estimation.Rmd-->
</div>
</div>
<div id="differential-gene-expression-analysis" class="section level1" number="6">
<h1><span class="header-section-number">6</span> Differential Gene Expression Analysis</h1>
<div id="r-data-import" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> R data import</h2>
<p>Load the proper conda environments, then launch R from the DESeq2 directory.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate deseq2-r</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate visualization</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>You are now in the R environment. Load the DESeq2 library (accessible because we first activated the appropriate conda environment).</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&#39;DESeq2&#39;</span>)</span></code></pre></div>
<p>You shouldn’t need to change the working directory, but you can check it with:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code></pre></div>
<p>and change it with:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&#39;/home/$USER/DGE_Virtual/DESEQ2/&#39;</span>)</span></code></pre></div>
<p>Load the table we made with featureCounts into a variable, then check the dimensions.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sampleCounts <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;DESEQ2_matrix.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="dv">1</span>, <span class="at">check.names=</span><span class="cn">FALSE</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sampleCounts)</span></code></pre></div>
</div>
<div id="format-metadata" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Format metadata</h2>
<p>Specify sample conditions. As the experimental design only incorporates two levels of a single factor, our table will be very simple. Nonetheless, it is important to create this accurately, as it is what the software will use to make comparisons.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sampleCondition</span> <span class="op">&lt;</span>- as.factor<span class="er">(</span><span class="ex">c</span><span class="er">(</span><span class="ex">rep</span><span class="er">(</span><span class="st">&quot;control&quot;</span><span class="ex">,3</span><span class="kw">)</span><span class="ex">,rep</span><span class="er">(</span><span class="st">&quot;stress&quot;</span><span class="kw">)))</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="ex">sampleTable</span> <span class="op">&lt;</span>- data.frame<span class="er">(</span><span class="va">condition</span><span class="op">=</span>sampleCondition<span class="kw">)</span></span></code></pre></div>
<p>Prepare the DESeq2-specific object.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>DESeqDataset <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData=</span>sampleCounts, <span class="at">colData=</span>sampleTable, <span class="at">design=</span><span class="sc">~</span>condition)</span></code></pre></div>
<p>Optional step 1: Count genes with totals &gt; 0 across all samples.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>GeneCounts <span class="ot">&lt;-</span> <span class="fu">counts</span>(DESeqDataset)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>idx.nz <span class="ot">&lt;-</span> <span class="fu">apply</span>(GeneCounts,<span class="dv">1</span>,<span class="cf">function</span>(x){<span class="fu">all</span>(x<span class="sc">&gt;</span><span class="dv">0</span>)})</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(idx.nz)</span></code></pre></div>
<p>Optional step 2: Generate library size normalization factors.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>DESeqSF <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(DESeqDataset)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sizeFactors</span>(DESeqSF)</span></code></pre></div>
</div>
<div id="normalization" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Normalization</h2>
<p>Load the library for plotting densities.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;geneplotter&quot;</span>)</span></code></pre></div>
<p>We “declare” the .pdf first, then run the plot command.
dev.off() is necessary to finish writing to the .pdf</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;count_density_prenorm.pdf&quot;</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">multiecdf</span>(<span class="fu">counts</span>(DESeqSF, <span class="at">normalized=</span>F)[idx.nz,],<span class="at">main=</span><span class="st">&quot;ECDF pre-normalized&quot;</span>, <span class="at">xlab=</span><span class="st">&quot;count&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;portion of genes&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1000</span>))</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/count_density_prenorm-1.png" style="width:100.0%" alt="" />
<p class="caption">prenorm</p>
</div>
<p>Now we’ll make a normalized one for comparison.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;count_density_postnorm.pdf&quot;</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">multiecdf</span>(<span class="fu">counts</span>(DESeqSF, <span class="at">normalized=</span>T)[idx.nz,],<span class="at">main=</span><span class="st">&quot;ECDF post-normalized&quot;</span>, <span class="at">xlab=</span><span class="st">&quot;count&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;portion of genes&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1000</span>))</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/count_density_postnorm-1.png" style="width:100.0%" alt="" />
<p class="caption">postnorm</p>
</div>
<p>Let’s generate similar multidensity plots.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;multidensity_prenorm.pdf&quot;</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">multidensity</span>(<span class="fu">counts</span>(DESeqDataset, <span class="at">normalized=</span>F)[idx.nz,],<span class="at">xlab=</span><span class="st">&quot;mean counts&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1000</span>))</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/multidensity_prenorm-1.png" style="width:100.0%" alt="" />
<p class="caption">multiprenorm</p>
</div>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;multidensity_postnorm.pdf&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">multidensity</span>(<span class="fu">counts</span>(DESeqSF, <span class="at">normalized=</span>T)[idx.nz,],<span class="at">xlab=</span><span class="st">&quot;mean counts&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1000</span>))</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="Figures/multidensity_postnorm-1.png" style="width:100.0%" alt="multipostnorm" />
## DESeq2</p>
<p>Designate reference level for each factor</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(DESeqDataset)<span class="sc">$</span>condition <span class="ot">&lt;-</span> <span class="fu">relevel</span>(DESeqDataset<span class="sc">$</span>condition, ref <span class="st">&quot;control&quot;</span>)</span></code></pre></div>
<p>Run DESeq2</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(DESeqDataset)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code></pre></div>
<p>The “results” function pulls the last comparison in the results list. Provide the an independent filtering variable equivalent to the study’s false discovery rate.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>stress_from_control <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">alpha=</span><span class="fl">0.05</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(stress_from_control)</span></code></pre></div>
<p>There are a few ways to extract comparisons from the results object.
What is different about this method?</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>control_from_stress <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span><span class="fu">c</span>(<span class="st">&quot;condition&quot;</span>,<span class="st">&quot;control&quot;</span>,<span class="st">&quot;stress&quot;</span>), <span class="at">alpha=</span><span class="fl">0.05</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(control_from_stress)</span></code></pre></div>
<p>Sort our results table, and export a file with counts appended to it.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>stress_from_control_sorted <span class="ot">&lt;-</span> stress_from_control[<span class="fu">order</span>(stress_from_control<span class="sc">$</span>padj),]</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">merge</span>(<span class="fu">as.data.frame</span>(stress_from_control_sorted),<span class="fu">as.data.frame</span>(<span class="fu">counts</span>(dds,<span class="at">normalized=</span><span class="cn">TRUE</span>)), <span class="at">by=</span><span class="st">&quot;row.names&quot;</span>), <span class="at">row.names=</span><span class="cn">TRUE</span>, <span class="at">quote=</span><span class="cn">FALSE</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">sort=</span><span class="cn">FALSE</span>, <span class="at">file=</span><span class="st">&quot;SfromC.tsv&quot;</span>)</span></code></pre></div>
</div>
<div id="heatmap" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Heatmap</h2>
<p>Load libraries for heatmap and subset our results.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;gplots&quot;</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>sigIndices <span class="ot">&lt;-</span> <span class="fu">which</span>(stress_from_control<span class="sc">$</span>padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>sigIndices</span></code></pre></div>
<p>Create object to feed into the plot.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>plotCounts <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds[sigIndices,])</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(plotCounts) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;C71&quot;</span>,<span class="st">&quot;C72&quot;</span>,<span class="st">&quot;C73&quot;</span>,<span class="st">&quot;S74&quot;</span>,<span class="st">&quot;S75&quot;</span>,<span class="st">&quot;S76&quot;</span>) </span></code></pre></div>
<p>Make it!</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span> (<span class="st">&quot;heatmapSigs.pdf&quot;</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>(plotCounts,  <span class="at">col=</span>hmcol, <span class="at">cexRow=</span><span class="fl">0.8</span>,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">key.title =</span> <span class="cn">NA</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, <span class="at">dendrogram=</span><span class="st">&quot;col&quot;</span>, </span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info=</span><span class="st">&#39;histogram&#39;</span>, <span class="at">cexCol=</span><span class="fl">0.6</span>, <span class="at">margin=</span><span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">10</span>))</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/heatmapSigs-1.png" style="width:100.0%" alt="" />
<p class="caption">workflow</p>
</div>
<p>Camk2n1 is skewing the color palette. Let’s make a version omitting it.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span> (<span class="st">&quot;heatmapSigsMost.pdf&quot;</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>(plotCounts[<span class="sc">-</span><span class="dv">7</span>,],  <span class="at">col=</span>hmcol, <span class="at">cexRow=</span><span class="fl">0.8</span>,</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">key.title =</span> <span class="cn">NA</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, <span class="at">dendrogram=</span><span class="st">&quot;col&quot;</span>, </span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info=</span><span class="st">&#39;none&#39;</span>, <span class="at">cexCol=</span><span class="fl">0.6</span>, <span class="at">margin=</span><span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">10</span>))</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/heatmapSigsMost-1.png" style="width:100.0%" alt="" />
<p class="caption">workflow</p>
</div>
<p>Principal Component Analysis now.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlogTransformation</span>(dds,<span class="at">blind=</span><span class="cn">TRUE</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;PCA.allsamples2.pdf&quot;</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(rld,<span class="at">intgroup=</span><span class="fu">c</span>(<span class="st">&quot;condition&quot;</span>))</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/PCA.allsamples2-1.png" style="width:100.0%" alt="" />
<p class="caption">PCA</p>
</div>
</div>
<div id="gene-count-plots" class="section level2" number="6.5">
<h2><span class="header-section-number">6.5</span> Gene count plots</h2>
<p>This last command will put a lot together. Let’s discuss:
loop iterations
aesthetic layers
x-axis labels and jittering</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(sigIndices)) {  </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(stress_from_control_sorted)[k]</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  pval <span class="ot">&lt;-</span> <span class="fu">format</span>(stress_from_control_sorted<span class="sc">$</span>padj[k], <span class="at">digits=</span><span class="dv">4</span>)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  d <span class="ot">&lt;-</span> <span class="fu">plotCounts</span>(dds, <span class="at">gene=</span>gene, <span class="co"># replaced = TRUE, </span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">&quot;condition&quot;</span>), </span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">returnData =</span> <span class="cn">TRUE</span>)</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(d, <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>count,</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>                     <span class="at">color=</span>condition)) <span class="sc">+</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">position=</span><span class="fu">position_jitter</span>(<span class="at">w=</span><span class="fl">0.2</span>, <span class="at">h=</span><span class="dv">0</span>), </span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha=</span><span class="dv">100</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">size=</span><span class="dv">3</span>, <span class="at">stroke=</span><span class="fl">1.25</span>) <span class="sc">+</span> </span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span><span class="fu">colnames</span>(plotCounts)), <span class="at">hjust=</span><span class="sc">+</span><span class="fl">1.2</span>, <span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;normalized counts&quot;</span>) <span class="sc">+</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(gene), </span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">subtitle =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsave</span>(<span class="fu">paste</span>(gene,<span class="st">&quot;.pdf&quot;</span>,<span class="at">sep=</span><span class="st">&quot;&quot;</span>))</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/Egr1.png" style="width:100.0%" alt="" />
<p class="caption">workflow</p>
</div>
<!--chapter:end:07DGE.Rmd-->
</div>
</div>
<div id="additional-visualization" class="section level1" number="7">
<h1><span class="header-section-number">7</span> Additional visualization</h1>
<div id="graphing-with-ggplot2" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Graphing with ggplot2</h2>
<p>Handy gglot2 references:</p>
<p>• ggplot2 gallery</p>
<p>– <a href="http://www.r-graph-gallery.com/portfolio/ggplot2-package/" class="uri">http://www.r-graph-gallery.com/portfolio/ggplot2-package/</a></p>
<p>• ggplot2 cheatsheets</p>
<p>– <a href="http://ggplot2.tidyverse.org/reference/" class="uri">http://ggplot2.tidyverse.org/reference/</a></p>
<p>• ggplot2 documentation</p>
<p>– <a href="https://cran.r-project.org/web/packages/ggplot2/ggplot2.pdf" class="uri">https://cran.r-project.org/web/packages/ggplot2/ggplot2.pdf</a></p>
</div>
<div id="data-and-setup" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Data and setup</h2>
<p>Reconnect to the DGE analysis screen.</p>
<p>If you haven’t already done so, activate the environment, open R, and load the ggplot2 library.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>source activate visualization</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>R</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;SfromCcomplete.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(expr)</span></code></pre></div>
</div>
<div id="scatter-plots" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Scatter plots</h2>
<p>Let’s compare replicates 1 and 2 for the control.</p>
<p>Treatment (Brain water avoidance stress, 1 hr)
Rep 1: SRR23869776
Rep 2: SRR23869775
Rep 3: SRR23869774</p>
<p>Control
Rep 1: SRR23869773
Rep 2: SRR23869772
Rep 3: SRR23869771</p>
<p>The aesthetics (aes) map variables onto parts of the plot. In this case, the x and y axes. We’ll put the plot into a variable then print the variable. Later on this will be valuable because we can add onto plots without regenerating them. We’ll print the plot to a pdf file. Alpha makes the points transparent so that it is easier to see when you have a lot of points in the same place on the graph.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;expression_replicates.pdf&quot;</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>scatexpr <span class="ot">=</span> <span class="fu">ggplot</span>(expr, <span class="fu">aes</span>(<span class="at">x=</span>SRR23869772, <span class="at">y=</span>SRR23869773)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.2</span>)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>scatexpr</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/expression_replicates-1.png" style="width:100.0%" alt="" />
<p class="caption">Scatterplot Control Replicates 1 and 2</p>
</div>
<p>Now let’s add some color based on the padj values. Because padj is on a continuous scale, ggplot2 will give us a scale for color. If we had used a qualitative variable it would give us discrete colors.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;expression_replicates2.pdf&quot;</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>scatexpr2 <span class="ot">=</span> <span class="fu">ggplot</span>(expr, <span class="fu">aes</span>(<span class="at">x=</span>SRR23869772, <span class="at">y=</span>SRR23869773, <span class="at">color=</span>padj)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.2</span>)</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>scatexpr2</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/expression_replicates2-1.png" style="width:100.0%" alt="" />
<p class="caption">Scatterplot Control Replicates 1 and 2 (Color)</p>
</div>
<p>Exercise: Compare two other reps.</p>
<p>Let’s compare the averages of the treatment to the averages of the controls. We can add new columns and calculate the means. Note that the dollar sign after the data frame name signifies a column. The square brackets after the data frame name signifies rows and columns. Since we left the row blank it will calculate the averages row by row across the designated columns.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>expr<span class="sc">$</span>ControlAverage <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(expr[ , <span class="fu">c</span>(<span class="dv">8</span><span class="sc">:</span><span class="dv">10</span>)])</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>expr<span class="sc">$</span>TreatmentAverage <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(expr[ , <span class="fu">c</span>(<span class="dv">11</span><span class="sc">:</span><span class="dv">13</span>)])</span></code></pre></div>
<p>Exercise: Create a scatterplot between the treatment and control averages that you just made.</p>
<div class="figure">
<img src="Figures/expression_replicates_averages-1.png" style="width:100.0%" alt="" />
<p class="caption">Scatterplot Treatment vs Control Averages</p>
</div>
</div>
<div id="volcano-plot" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Volcano plot</h2>
<p>We can even make a volcano plot in ggplot2, which is really just a specialized scatter plot. The x axis is the log2 of the fold change (Fold_Change). This transformation creates the symetrical “V” shape of the Volcano Plot. The y axis is the negative log10 of the p value (padj), which makes lower number (more significant) higher on the plot. We’ll also add a line at p=0.05 (on the graph this will be -log10(p)=1.3). Those above the line are significant. We’ll color by padj. This time, we’ll add layers to the base graph we put in the “volcano” variable rather than redoing the graph from scratch.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>neglog10padj <span class="ot">=</span> <span class="sc">-</span><span class="fu">log10</span>(expr<span class="sc">$</span>padj)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>volcanoDF <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">lg2FC =</span> expr<span class="sc">$</span>log2FoldChange, <span class="at">negLG =</span> neglog10padj, <span class="at">padj=</span> expr<span class="sc">$</span>padj)</span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span> (<span class="st">&quot;volcano.pdf&quot;</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>volcano1 <span class="ot">=</span> <span class="fu">ggplot</span>(volcanoDF, <span class="fu">aes</span>(<span class="at">x=</span>lg2FC, <span class="at">y=</span>negLG, <span class="at">color =</span> padj)) <span class="sc">+</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept=</span><span class="fl">1.3</span>, <span class="at">color=</span><span class="st">&#39;gray&#39;</span>) <span class="sc">+</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;log2(Fold Change)&quot;</span>) <span class="sc">+</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;-log10(Adjusted p value)&quot;</span>) <span class="sc">+</span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Volcano Plot&quot;</span>)</span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>volcano1</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/volcano-1.png" style="width:100.0%" alt="" />
<p class="caption">Volcano Plot</p>
</div>
<p>And now, let’s play with the colors. We will color by the log2(Fold Change) instead of padj. We’ll use a more recent color scheme, viridis (purple to yellow) which is gaining traction in gene expression circles. According to the developers (<a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html" class="uri">https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html</a>), it was developed to be:</p>
<p>• Colorful, spanning as wide a palette as possible so as to make differences easy to see</p>
<p>• Perceptually uniform, meaning that values close to each other have similar-appearing colors and values far away from each other have more different-appearing colors, consistently across the range of values</p>
<p>• Robust to colorblindness, so that the above properties hold true for people with common forms of colorblindness, as well as in grey scale printing, and</p>
<p>• Pretty, oh so pretty</p>
<p>scale_color_gradient2(low = “blue”, mid = “white”, high = “red”)</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span> (<span class="st">&quot;volcanoimproved.pdf&quot;</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>volcano_viridis <span class="ot">=</span> <span class="fu">ggplot</span>(volcanoDF, <span class="fu">aes</span>(<span class="at">x=</span>lg2FC, <span class="at">y=</span>negLG, <span class="at">color =</span> lg2FC)) <span class="sc">+</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept=</span><span class="fl">1.3</span>, <span class="at">color=</span><span class="st">&#39;gray&#39;</span>) <span class="sc">+</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;log2(Fold Change)&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;-log10(Adjusted p value)&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Volcano Plot&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>()</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>volcano_viridis</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/volcanoimproved-1.png" style="width:100.0%" alt="" />
<p class="caption">Volcano Plot viridis</p>
</div>
</div>
<div id="boxplots" class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> Boxplots</h2>
<p>In expression data, you often want to print a boxplot of some of your interesting genes based on differential expression and/or the biology of the gene. They show the mean and the interquartile range. The whiskers go to the farthest distal value from the nearest quartile that is no farther than 1.5 * IQR beyond the nearest quartile. Any outliers are plotted as individual dots.</p>
<p>Let’s take a closer look at that gene that stood out before (Camk2n1).</p>
<p>First, get the row for the Camk2n1 gene using the subset command.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>df_camk2n1 <span class="ot">=</span> <span class="fu">subset</span>(expr, Row.names <span class="sc">==</span> <span class="st">&quot;Camk2n1&quot;</span>)</span></code></pre></div>
<p>Now grab the columns (8-13) with the replicate data and transpose them (t). Note that we are using the same variable name. The transpose command turns it into a matrix but we need a data frame, so we’ll wrap everything in the as.data.frame command to keep it as a data frame.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>df_camk2n1 <span class="ot">=</span> <span class="fu">as.data.frame</span>( <span class="fu">t</span>(df_camk2n1[,<span class="dv">8</span><span class="sc">:</span><span class="dv">13</span>]) )</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at it</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>df_camk2n1</span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">30</span> </span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>SRR23869771 <span class="fl">96840.79</span> </span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>SRR23869772 <span class="fl">98702.66</span> </span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>SRR23869773 <span class="fl">93495.25</span> </span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>SRR23869774 <span class="fl">62737.18</span> </span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>SRR23869775 <span class="fl">83620.25</span> </span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>SRR23869776 <span class="fl">61241.84</span></span></code></pre></div>
<p>Right now the column header is 30, referring to the original row number for Camk2n1, rather than the gene name. Let’s fix that.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df_camk2n1) <span class="ot">=</span> <span class="st">&quot;Camk2n1&quot;</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>df_camk2n1</span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>             Camk2n1 </span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>SRR23869771 <span class="fl">96840.79</span> </span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>SRR23869772 <span class="fl">98702.66</span> </span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>SRR23869773 <span class="fl">93495.25</span> </span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>SRR23869774 <span class="fl">62737.18</span> </span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>SRR23869775 <span class="fl">83620.25</span> </span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>SRR23869776 <span class="fl">61241.84</span></span></code></pre></div>
<p>That’s better. Now we need a new column with the treatment condition.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>Treatment<span class="ot">=</span><span class="fu">c</span>(<span class="st">&quot;Control&quot;</span>,<span class="st">&quot;Control&quot;</span>,<span class="st">&quot;Control&quot;</span>,<span class="st">&quot;Stress&quot;</span>,<span class="st">&quot;Stress&quot;</span>,<span class="st">&quot;Stress&quot;</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>df_camk2n1</span></code></pre></div>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>             Camk2n1 Treatment</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>SRR23869771 <span class="fl">96840.79</span>   Control</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>SRR23869772 <span class="fl">98702.66</span>   Control</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>SRR23869773 <span class="fl">93495.25</span>   Control</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>SRR23869774 <span class="fl">62737.18</span>    Stress</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>SRR23869775 <span class="fl">83620.25</span>    Stress</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>SRR23869776 <span class="fl">61241.84</span>    Stress</span></code></pre></div>
<p>Now for the boxplot.</p>
<p>The first part, including the aesthetics, takes the same form we saw in the scatterplots. Instead of adding a geom_point() layer, we’ll add a geom_boxplot() layer. We’ll color by Treatment.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;boxplot.pdf&quot;</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>boxplot <span class="ot">=</span> <span class="fu">ggplot</span>(df_camk2n1, <span class="fu">aes</span>(<span class="at">x=</span>Treatment, <span class="at">y=</span>Camk2n1, <span class="at">fill=</span>Treatment)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>()</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>boxplot</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/boxplot-1.png" style="width:100.0%" alt="" />
<p class="caption">Boxplots</p>
</div>
<p>Excercise: Try making boxplots with other genes.</p>
</div>
<div id="violin-plots" class="section level2" number="7.6">
<h2><span class="header-section-number">7.6</span> Violin plots</h2>
<p>Violin Plots are similar to box plots but they show the distribution of the data a little bit better. They use rotated density plots to show the probability of each datapoint being a particular value. It is the same as a boxplot except using geom_violin instead of geom_boxplot.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;violinplot.pdf&quot;</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>violin <span class="ot">=</span> <span class="fu">ggplot</span>(df_camk2n1,</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x=</span>Treatment, <span class="at">y=</span>Camk2n1, <span class="at">fill=</span>Treatment)) <span class="sc">+</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>()</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>violin</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/violinplot-1.png" style="width:100.0%" alt="" />
<p class="caption">Violin plot</p>
</div>
<p>You can even use both a violin_plot layer and a box_plot layer. Note that the order matters as one writes on top of the other. Usually you put a skinny boxplot (using the “width” command) inside a violin plot. We’ll make the boxplot somewhat skinny and use “alpha” to make it transparent so you can still see the violin plot through it.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;violinbox.pdf&quot;</span>)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>violin <span class="ot">=</span> <span class="fu">ggplot</span>(df_camk2n1, <span class="fu">aes</span>(<span class="at">x=</span>Treatment, <span class="at">y=</span>Camk2n1, <span class="at">fill=</span>Treatment)) <span class="sc">+</span> <span class="fu">geom_violin</span>() <span class="sc">+</span>  <span class="fu">geom_boxplot</span>(<span class="at">width=</span><span class="fl">0.1</span>,<span class="at">alpha=</span><span class="fl">0.3</span>)</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>violin</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<div class="figure">
<img src="Figures/violinbox-1.png" style="width:100.0%" alt="" />
<p class="caption">ViolinBox plot</p>
</div>
</div>
<div id="interactive-plots" class="section level2" number="7.7">
<h2><span class="header-section-number">7.7</span> Interactive Plots</h2>
<p>The plotly library can make ggplot2 plots interactive. Since we have all of our ggplot2 plots in variables we can use those variables to make the plot interactive. We’ll use the plotly library as well as the htmlwidgets library which allows us to avoid pandoc installation and setting of browsers. We’ll use the scatterplot as an example. The scatterplot is in the variable “scatexpr2”.</p>
<p>Load the libraries.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plotly)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(htmlwidgets)</span></code></pre></div>
<p>Apply ggplotly to the variable that contains the ggplot2 plot.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>scatint <span class="ot">=</span> <span class="fu">ggplotly</span>(scatexpr2)</span></code></pre></div>
<p>Save it as an HTML file with an accompanying library folder (“lib”)</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveWidget</span>(scatint, <span class="st">&quot;scatterplot.html&quot;</span>, <span class="at">selfcontained =</span> F, <span class="at">libdir =</span> <span class="st">&quot;lib&quot;</span>)</span></code></pre></div>
<p>You need to copy both the html file and the “lib” directory to the same directory on your computer. Because “lib” is a directory, you need to use the recursive flag (-r) if you are using scp. You should be able to double-click on the html file to open it.</p>
<div class="figure">
<img src="Figures/InteractiveScatter.png" style="width:100.0%" alt="" />
<p class="caption">Interactive Scatterplot</p>
</div>
<p>Try zooming in on the blue dots in the bottom left corner by selecting the magnifying glass and making a box around them.</p>
<p>Exercise: Now make an interactive plot for the other ggplot plots. If more libraries are needed for the html, they will be added to the library folder, so I recommend that you copy the library folder over each time.</p>
<ol style="list-style-type: decimal">
<li><p>Volcano plot (“volcano_viridis” variable)</p></li>
<li><p>Box plot (“boxplot” variable)</p></li>
<li><p>Violin plot (“violin” variable)</p></li>
</ol>
<p>We can also use plotly directly for interactive plots that are not from ggplot2. We’ll start with a 3D scatterplot. We’ll use the same expr data frame to look at all 3 reps for the Control.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>scatter3d<span class="ot">=</span><span class="fu">plot_ly</span>(</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span>expr<span class="sc">$</span>SRR23869773,</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y=</span>expr<span class="sc">$</span>SRR23869772,</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">z=</span>expr<span class="sc">$</span>SRR23869771,</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">type=</span><span class="st">&quot;scatter3d&quot;</span>, <span class="at">mode=</span><span class="st">&quot;markers&quot;</span>, <span class="at">color=</span>expr<span class="sc">$</span>padj)</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="fu">saveWidget</span>(scatter3d, <span class="st">&quot;scatter3d.html&quot;</span>, <span class="at">selfcontained =</span> F, <span class="at">libdir =</span> <span class="st">&quot;lib&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Figures/Scatter3D.png" style="width:100.0%" alt="" />
<p class="caption">Interactive 3D Scatterplot</p>
</div>
<p>Interact with the plot and see what you can do with it.</p>
<p>Exercise: Now make an interactive 3D scatterplot for the 3 replicates for the Treatment.</p>
<p>We’ll also make an interactive heatmap. We’ll use Heatmaply for this. Heatmaply can be used for either static or interactive heatmaps.</p>
<p>We’ll need a matrix of genes x samples. First, we’ll make the genes the rownames, rather than the default of row numbers. Note that the genes are in the Row.names column.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the genes the rownames</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr) <span class="ot">=</span> expr<span class="sc">$</span>Row.names</span></code></pre></div>
<p>We’ll grab the significant genes (padj&lt;=0.05) and just the rep data (columns 8-13). We’ll do it all in one command but you can break it into two commands if that is easier.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>expr2 <span class="ot">=</span> <span class="fu">subset</span>(expr, padj<span class="sc">&lt;=</span><span class="fl">0.05</span>)[,<span class="dv">8</span><span class="sc">:</span><span class="dv">13</span>]</span></code></pre></div>
<p>Check the data frame using the head command.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(expr2)</span></code></pre></div>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>        SRR23869771 SRR23869772 SRR23869773 SRR23869774 SRR23869775 SRR23869776</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>Gm5537    <span class="fl">60.357391</span>   <span class="fl">57.400286</span>    <span class="fl">66.32549</span>    <span class="fl">5.891184</span>    <span class="fl">7.326479</span>    <span class="fl">6.761291</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>Gm10154   <span class="fl">32.825950</span>   <span class="fl">47.833571</span>    <span class="fl">42.63781</span>  <span class="fl">142.370279</span>  <span class="fl">123.503508</span>   <span class="fl">99.487567</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>Ddc       <span class="fl">83.653227</span>  <span class="fl">103.320514</span>    <span class="fl">95.69821</span>  <span class="fl">461.476078</span>  <span class="fl">142.343026</span>  <span class="fl">279.144727</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>Gm10231  <span class="fl">188.484485</span>   <span class="fl">76.533714</span>    <span class="fl">38.84779</span>   <span class="fl">20.619144</span>   <span class="fl">24.072718</span>   <span class="fl">24.147468</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>Zfhx3    <span class="fl">728.524303</span>  <span class="fl">495.555799</span>   <span class="fl">480.38605</span> <span class="fl">1345.153674</span>  <span class="fl">723.228170</span> <span class="fl">1165.839741</span></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>Mab21l1    <span class="fl">7.412311</span>    <span class="fl">7.653371</span>    <span class="fl">13.26510</span>   <span class="fl">81.494712</span>   <span class="fl">12.559679</span>   <span class="fl">88.862681</span></span></code></pre></div>
<p>Change the data frame into a matrix.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>expr2matrix <span class="ot">=</span> <span class="fu">as.matrix</span>(expr2)</span></code></pre></div>
<p>Now, let’s make the heatmap.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(heatmaply)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>h<span class="ot">=</span><span class="fu">heatmaply</span>(expr2matrix)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="fu">saveWidget</span>(h, <span class="st">&quot;heatmap.html&quot;</span>, <span class="at">selfcontained =</span> F, <span class="at">libdir =</span> <span class="st">&quot;lib&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Figures/Heatmap3D.png" style="width:100.0%" alt="" />
<p class="caption">Interactive Heatmap</p>
</div>
</div>
<div id="homework-optional" class="section level2" number="7.8">
<h2><span class="header-section-number">7.8</span> Homework (optional)</h2>
<p>Use ggplot2 to plot other types of graphs</p>
<p>(<a href="http://r-statistics.co/Top50-Ggplot2-Visualizations-MasterList-R-Code.html" class="uri">http://r-statistics.co/Top50-Ggplot2-Visualizations-MasterList-R-Code.html</a>)</p>
<p>Try to make some of them into interactive graphs using plotly</p>
<!--chapter:end:08Additional_Visualization.Rmd-->
</div>
</div>
<div id="pathway-enrichment-analysis" class="section level1" number="8">
<h1><span class="header-section-number">8</span> Pathway Enrichment Analysis</h1>
<div id="parsing" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Parsing</h2>
<p>Let’s fix up the file headers</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&#39;s/\/home\/elavelle\/DGE_workshop\/alignments\///g&#39;</span> SfromC.tsv <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/\.sorted\.bam//g&#39;</span> <span class="op">&gt;</span> SfromCsed.tsv</span></code></pre></div>
<p>Use a couple options to enable better viewing.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&#39;s/\/home\/elavelle\/DGE_workshop\/alignments\///g&#39;</span> SfromC.tsv <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/\.sorted\.bam//g&#39;</span> <span class="op">&gt;</span> SfromCsed.tsv</span></code></pre></div>
<p>Let’s filter out genes we are missing data for.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">&#39;$6 != &quot;NA&quot; {print $0}&#39;</span> SfromCsed.tsv  <span class="op">&gt;</span> SfromCfiltered.tsv</span></code></pre></div>
<p>We’ve made a good results table. But the researchers didn’t report DEGs based on adjusted p-value. To compare our results, let’s sort by regular p-value.</p>
<p>Preserve the header, run a generalized sort from the second line on, then concatenate the two back together.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-1</span> SfromCfiltered.tsv <span class="op">&gt;</span> header</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n+2</span> SfromCfiltered.tsv <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-g</span> <span class="at">-k6</span> <span class="op">&gt;</span> SfromCsorted.tsv</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> header SfromCsorted.tsv <span class="op">&gt;</span> SfromCcomplete.tsv</span></code></pre></div>
<p>Download the file.</p>
</div>
<div id="cytoscape-and-cluego" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Cytoscape and ClueGO</h2>
<p>Open the app, and load the ClueGO plug-in: Apps –&gt; App Manager
Open the results table in a spreadsheet, copy the DEGs, an paste into clueGO.</p>
<p>Selection in the “Visual Style” box sets the factors that will determine node properties. Keep the default selection, “Groups”.</p>
<p>Pick ontologies and set differing node shapes.</p>
<p>Clicking on the “Evidence” button will display a decision tree clarifying evidence codes. Keep “All” checked for now.</p>
<p>Check the “Show only Pathways with pV &lt;=” box. If you want the details on how p-values are calculated for GO terms, click the “?” icon.</p>
<p>Expand the “Statistical Options” menu. In the first box, change the selection from the default (“Enrichment/Depletion”) to “Enrichment. A left side to the hypergeometric test will include significantly depleted pathways.</p>
<p>Run the analysis! It will take several minutes. When it’s finished, look to the bottom right of the screen. Select the “ClueGO Results (Cluster #1)” tab, and click the indicated icon to save ClueGO results tables to your machine.</p>
<p>Open the new directory. Among the several files, we can find a .png with GO terms represented with a horizontal bar graph as they are shown in the “Cluster #1” tab of in the software. Adjacent bars of the same color are terms that belong to the same group.</p>
<p>The most generally useful file is “NodeAttributeTables.txt”. The fields here indicate annotation source, p-values, and gene composition and for each gene. Note the “GOLevels” column and the “Overview Term” column on the far right. Let’s return to Cytoscape to clarify what these mean.</p>
<p>Each GO term has one or more associated levels. Only those with a level within the values in the “GO Tree Interval” have the potential to appear in the results. The boundaries of this window can be changed directly, or by dragging the “Network Specificity” bar.</p>
</div>
<div id="additional-options" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Additional options</h2>
<p>Let’s inspect some additional options.</p>
<p>It’s typically a good idea to check the “Use GO Term Fusion” box. This will improve runtime and simplify the results by eliminating redundancy.</p>
<p>Below the “GO Tree Interval” box, there are two others. The “GO Term/Pathway Selection” box will set the criterion for what is a qualifying GO term.</p>
<p>The “GO Term/Pathway Network Connectivity” scroll tab will set the kappa score threshold for what the edges that appear between nodes. Kappa score is a measure of relation any and all GO terms based on the number of associated genes they have in common. For more details regarding this calculation, click on the “?” icon.</p>
<p>Expand the “Grouping Options” box. Recall the variable (True/False) of the last column (Overview Term) in the node attribute table file we looked at. Those with a “True” value (as well as a check mark next to the GO id in the “Clue GO Results” tab in Cytoscape) are the leading term in the group, only meaning it is the group with the lowest p-value.</p>
<p>The criterion to determine the leading term can be changed via the “Leading Group Term based on” selection. The threshold of commonality for what constitutes a group can also be adjusted, if so chosen.</p>
<p>Let’s run the analysis again, selecting the KEGG ontology instead of the GO ontologies.</p>
</div>
<div id="cluepedia" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> CluePedia</h2>
<p>Select the “CluePedia” tab, and click the network icon. This will add miniature gene nodes within the edges between nodes. As of now, however, it is not apparent which of them are upregulated and which are downregulated.</p>
<p>Navigate to File –&gt; Import –&gt; Table from File, and select the document from which the gene list was originally input. In the new window, chose “To selected networks only”. Under the “Key Column for Network” menu, pick “Input Gene ID”, then “OK”.</p>
<p>In the control panel, on the left side of Cytoscape, select ”Style”. For column, choose “Log2Fold change” and for mapping type, select “Continuous Mapping” and click the spectrum map.</p>
<p>In the window that appears, drag the handle position to 0 and change the color to grey or white. Then, drag the margins in to about a log2FoldChange of 2 and change their color. Change the arrows on the extremes to the same color.</p>
<p>You can save the session by clicking the icon at the top right of the app, which will allow you to return to the visualization and settings of the analysis in another session.</p>
<!--chapter:end:09PE.Rmd-->
</div>
</div>
<div id="de-chapter-10" class="section level1" number="9">
<h1><span class="header-section-number">9</span> DE Chapter 10</h1>
<!--chapter:end:10.Rmd-->
</div>
<!--bookdown:body:end-->
  </main>

  <div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page">
      <h2>On this page</h2>
      <div id="book-on-this-page"></div>

      <div class="book-extra">
        <ul class="list-unstyled">
          <li><a id="book-source" href="#">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="#">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
      </div>
    </nav>
  </div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5">
  <div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Differential Expression Workshop NCGR</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>


</body>

</html>
